{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hyun-Joon Yang\n",
    "### yanghyun@usc.edu\n",
    "### QBIO 401\n",
    "### Assignment 7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are going to build models to predict protein–DNA binding based on the data obtained from high-throughput in vitro protein–DNA binding assays.\n",
    "\n",
    "Download and install the following packages from Jupyter Notebook or other platforms:\n",
    "1. Bioconductor (refer the instructions on https://www.bioconductor.org/install/)\n",
    "2. The DNA shape prediction tool: DNAshapeR ( refer the instructions on https://www.bioconductor.org/packages/release/bioc/html/DNAshapeR.html)\n",
    "3. The machine learning tool: caret (refer the instructions on https://github.com/topepo/caret)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "#     install.packages(\"BiocManager\")\n",
    "\n",
    "# BiocManager::install(\"DNAshapeR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# install.packages('caret')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rm(list=ls())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicRanges\n",
      "Loading required package: stats4\n",
      "Loading required package: BiocGenerics\n",
      "Loading required package: parallel\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "The following objects are masked from 'package:parallel':\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Warning message:\n",
      "\"package 'S4Vectors' was built under R version 3.6.3\"\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Loading required package: IRanges\n",
      "Warning message:\n",
      "\"package 'IRanges' was built under R version 3.6.2\"\n",
      "Attaching package: 'IRanges'\n",
      "\n",
      "The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "Warning message:\n",
      "\"package 'GenomeInfoDb' was built under R version 3.6.3\"Warning message:\n",
      "\"package 'caret' was built under R version 3.6.3\"Loading required package: lattice\n",
      "Loading required package: ggplot2\n",
      "Registered S3 methods overwritten by 'ggplot2':\n",
      "  method         from \n",
      "  [.quosures     rlang\n",
      "  c.quosures     rlang\n",
      "  print.quosures rlang\n"
     ]
    }
   ],
   "source": [
    "library(DNAshapeR)\n",
    "library(caret)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. Use the DNAshapeR package to predict DNA shape for each sequence bound by transcription factor Mad, Max and Myc. The datasets were obtained from the in vitro gcPBM (genomic context protein binding microarray) assay. The sequence data in FASTA format (Mad.fa, Max.fa and Myc.fa) can be found on Blackboard. Use plotShape() or heatShape() functions of DNAshapeR to generate ensemble plots for the DNA shape parameters of minor groove width (MGW), propeller twist (ProT), Roll, and helix twist (HelT). Discuss what you find from the results [2pt]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing files......\n",
      "Record length: 36\n",
      "Record length: 35\n",
      "Record length: 36\n",
      "Record length: 35\n",
      "Record length: 36\n",
      "Done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing files......\n",
      "Record length: 36\n",
      "Record length: 35\n",
      "Record length: 36\n",
      "Record length: 35\n",
      "Record length: 36\n",
      "Done\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n",
      "Reading the input sequence......\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing files......\n",
      "Record length: 36\n",
      "Record length: 35\n",
      "Record length: 36\n",
      "Record length: 35\n",
      "Record length: 36\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "# names\n",
    "ns <- c('Mad', 'Max', 'Myc')\n",
    "# list for each shape output\n",
    "shapes <- list()\n",
    "# get shape for each file\n",
    "for (n in ns) {\n",
    "    fn <- paste(n,'.fa', sep='')\n",
    "    shapes[[length(shapes) + 1]] <- getShape(fn)\n",
    "}\n",
    "# add names\n",
    "names(shapes) <- ns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAD1BMVEUAAABGgrRNTU3Ozv//\n//8jJXzIAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO2djXqrKBRFTa7v/8wzaeIP\nckDQjRyTtb6Z2yRFwFU2ijHtMALAaYbeHQD4BggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEI\nIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACC\nBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEImIM0\nvOnZF5fgxQYvIR8TsxHUBODFBi9bhtW/Y/T418GLDV4i0AAgYEg+gfFvJTCyFIh5CcHLms+p\n3Uzn7jhj+PzPFBOCl4jgYgNaNjBgbPASsbr8jZYIBowNXiJWJjixi2DA2MxO8DLBxYYsXGyw\nYUm9BRUAAuaLDcy8FlzmtZmcIGZiubMBMTF4sWGNFEGQcuDFZh41eJkgSDnwYjMtCPAys7lp\nFTEBnMLYcO93xGB9gQku89oMm6+ACQABq1M7Zt4IvCTgQL2Fiw058GIzjCwdNxCkHHix4SJM\nBEHKgRcbghQxf7CPAWOAFxuCFLH6PBJvVFvgxeJzpQEvM6gAELC5swFCuCvehrvityxrJIjh\nYoMNXiKW+3gPOhla8WxW84tyOwcHTLOe48Wms5fgzoZqK2PDOenZquIXFwyY6i0KwYtNZy/B\nGqkkefVN1PMv+1TBPQeMKy8HL39fcDLYxcvpiw0NxDwe/zJPJZT2ejh8seHrvZSd8RxuoYbQ\nQx8vp/eryYB5ZJ5KaD8v4uWyFqIJpoeXYfXAyxrpbeIZPtVS3mu8JEvixSqBmGRJJ14eeLFx\n4cVhkBgwNoYXuZg7eiFICW4+YI6vw3fAi40PL8eD9OViViV9TDD/S8CLhY/x4u+INA2Y1QsM\nmBEvSbYeCNIbBoyN4WWUi/kOL12DdLS2bx8wXlp4KcCLgZPxspTIrHhy9+5dsRboOmAOrgSb\nBQkvIU7GyxA9MEsNySItBszfP8/VC31PYRq3UIbpRS7mfl6iIHUaL2VBGtN/GZMgabdL8DZA\nkCK8jJfSII2pW8OZebXbJSBICbyMlyVI++e8VwTJ34BxsRb43JdpeNHe6Xw3Lw6D1K6JKpYg\nBa/0C5KPFj77H3lRi7mbFz/jZYjKVe7pVWK6zLxHvDS548NZkNx48Rak5WPm5Sd3jQbMfAqz\nvLR9RUD5T3/7QN1CGc6C5MaLn4m3Zo10sIkaVgMmOOftdwqDl0RJJ17steP1XpytkfwNmHw1\n6U+i4+W3vBTt1/vt2N8SU1pL6q1qvNjVfauX4iD9mpjSWvBi19LLi7GaFHDXIL2+dhZTWktn\nL2IxX+OFIIVf/QcpVd1FXtwGKVXdt46XsiBl/kzQt4qpqab52nHe+3sE6TIvjsbLUmJOi10u\nddfq18+8eS+KFvbxGCQPXjwGaZj+kzdRg7+1gAsv6yDhZYWf8eIrSHcdMOtvt7jjAy8JviFI\nDBhRCwXgJYHDII0efim6wwGDl1RRvNSVaF3BGkdiimu54o6PjBetmO/xQpDCB+6DdMnbAhkv\nXoPU20u/NZKHu3kdiZlL5n670th/wPQKki8v8TDpe0Ryc86beUNJQmWvE8U9DBi8BI+cBKn3\nVZi1mGf0irCh2l4nZt6L7vhY7Tpe1jgaLwTpXPlL7vi4W5DGi7x4Gi+bNVKLJipwJGYu6cCL\nxyB58OJpvLi6audxwHhoAS8JCJINA8Ym60UqpqGXpnfC9B8v7xLDUL+bDcQYo6NvkO7gpUeQ\njnipa6EIT+PF7xFp+xJHpPXD+x2RGrTgabzcJ0gMmPVDvIxeg+TtUO1lLYCXRFFXXj6P41c0\n1Fz+fv3n+TJvpwGDl2RJT16cBanzG7KexKxK9vYS7DheFjyNl6Ig5S7SfKuYVcneA4YgJfA0\nXob1o+Qpb6aebxWzLtr5HfydICnF3MqLq/FS+sbB2QpKcDnzOmjBZ5ActHDDIDWsYIEg2RCk\nBATJhgFj8x1e9Hd8uJp4l4sNB/fw2wdM/V/SunrA3MRLdQsJEtf9817OGqo6IhX88NcFfmTA\nFHk52cIOeFnxSNy78PgQvGRsdIzaU7se71Q7nWHC4nixiyfKZybZk14eO0H6F7xkbXUI9RHp\nUBN5vM4wq7JpL+0GTJ2Xz9PT40XkZYgeHGsh5rWD8z6GO/t3x8e/sPDyIBRW3a5ojTS8C7Sa\nYUwxmRkmHknVzWrWAk0HzCod/6LvPP5tS49f62U7l+aCFG64jJf5iTEPlSC6ajdMBU+LMSbT\nlJh/8z9j+P1HsNGsJutm+zeuNSuYZkH67OJ7AGz2LPYyzhKWA1nhgPHu5RH+dMPz2r29W1SM\nc5oemxpTTVb3+sogGbvx2cfVk5LtV4UtNQ9jq82I8Ttggl0snzzXG9UMmP5e/iU7GklZ7+JY\nEaR1beuqM+1Ue9EH6V/Yo/n1+enqO5uf/XZ+3LKuchkw4TcTHNntXeqDZPQl112j50kmecaA\n2Upo7aVm7bgZLqmf22av1l+S5GVmm6w/UpcFqeL3lOXEGLsRP5mIz3k/DYQ1GHWmW032+iA1\nFxt2B8xyzrb8GI2pJeUlaKikXXOH9io/TTzxTv/s93C2YRWw1khhqcT58EZ+Wa8PlPiUK/49\nZXOX/h6ktaxfs6zVDpil8aAXNtcPmH+zhJIBnWHXi7Xbm59FctPud3xYs0nMZzeCIoaX3Xry\n31xRE6SGn3hMdffczFtswaBiP1Ve/sXPDu3AvpccO02KvTR7A9/Qd87LDjWXv5s1cZDOYqpK\ntr3jI+ROXoTb7XL7ILXi2azmF+X7hxe8FHg5HaRmc4yTavFSXz73Br6bHRBXu7qzoSx5x5u6\nZ7U5L784YObyO15SV3nd7IC4WkHz3ymmtK5fGzCldf2aF4J0sq5fGzCldf2al9OXv7vvQetq\nOYVJbJDxknkD388OaKtdX2w49Av/uu9B42pzXn5xwKzK58bLkHoD380OiKsNg3SoN98pZlU+\n5+XnBsyqPOPFKoeYZHm82OXxYpZ7aWnUmVuDFxu8BKACQABBAhAQXP4mVgZ4scFLgOBiw3eD\nFxu8hBCkHfBig5cQgrQDXmzwErK5RahjT7yCFxu8BKACQABBAhDwDlLFR2rNChRH+YtOFKr6\ni5dEYbyY5U7c7PG3YfoO6IqKzCrkrir7i5dM8QvayVRkVtHPy3LV7lAfhlEkZgh7E74s4dPJ\n2v7iJbEZXlItH/9Yn0zM0o3h7yf1eTy9omihvr94SWyIF7ND704da+/8Oe+wfjC8ezIsuzGc\nOJsIWzjUX7wktsVL1CEvM8ykYYzEnKp+XhrfdebFS1TNur7uXpYO1Tb88SgXM8wfOp2+LK8I\nWqjtL14Sm+HFLH64WamYYdv/ufZOAwYviY3wYnfo3PsCAjHrg3KDQ/VYLeZdGC/2NnhJFD/Z\nnvINtuDUNLgKI0HW36vawUuiIl9eWnsD+AkIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBI\nAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQg\ngCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAII\nEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCAB\nCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAA\nggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBI\nAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAALmIA1vevbFJXixwUvIx8RsBDUBeLHB\ny5Zh9e8YPf518GKDlwg0AAgYkk9g/FsJjCwFYl5C8LLmc2o307k7zhg+/zPFhOAlIrjYgJYN\nDBgbvESsLn+jJYIBY4OXiJUJTuwiGDA2sxO8THCxIQsXG2xYUm9BBYCA+WIDM68Fl3ltJieI\nmVjubEBMDF5sWCNFEKQceLGZRw1eJghSDrzYTAsCvMxsblpFTACnMDbc+x0xWF9ggsu8NsPm\nK2ACQMDq1I6ZNwIvCThQb+FiQw682AwjS8cNBCkHXmy4CBNBkHLgxYYgRcwf7GPAGODFhiBF\nrD6PxBvVFnix+FxpwMsMKgAEbO5sKCj7U3BXvE3RXfE/ZW1ZI+UK/ervRuFig82Ol18cL8t9\nvLmdznz4ZGjFs1nNL8rtHAxSs57fwcsPjpfgzoZcudSSu9mk82xV8YsLglS9RSH38PJz4yVY\nI2WTt/1mcVar+Ff5+kFuFyR/XvYuf//YeDl9sUEr5vGwDTz+R9lOaa+HwxcbpF4eDr3sp+Kn\nxsvp/VKKeST3P/2dY7RfBQtbeOx4EZq5lRdX42VYPdgpnbhCU9ejHOGgeIbf0k4y5b0++nas\nzsvfjue9yE5klF5+a7w4CtLfjqfFpM9vDnCvII1ZL+tvnuVOQfI1XlwFabzlgDnbwg7bXb+N\nl98aLxVBOtpEKb7ErEq6D5JOzI28OBsvfoLEgDGJ9hwvb5wdqf0GyfjmLw6Y/J5PL6vE3MeL\nt/HiJkhFA+b3grSTFILkZby4eR/J68zbuQW3QerdgreJdylx8AaOS9cCPQZMVy/z/u54UYm5\nixd/42WIHsibKCIeEU4GTF8vy+7iJSSKCkH643uCpLw5c/WW4o4XkZjbBKlwvPxckHZmXvV4\nucuAWe3tjheC1Hm8LEGqnUelt8UXB+n6AdNxLbDe130vCjENvWg/RuE3SO2aKMA6hQm/bz08\nwS2uTj3MIIUldgrUcgsv425SugVpiF4RNrHPYy8ovWberl7CPd0NkmTM3MHLi+IgXTbxfkrM\nR9zqQ+93z7w9vYQ7WhKk82bu4GV0eQZzfI1U3MQu9ryaWAtcOvOOHb1sdnPfiyJJN/AyJs5g\nOq8dHayREjNvJkjXzbzdWog+llYQJMEU497LC/sMJuVFkqQ7BOlRG6RrZ96LW5jf/9juY0mQ\nomAl65/YfvLNrZc1Ls9gegcpPWK+bebNflrzEWEUeX/NB+mxSmJcXfxs06s7BCkxGHIT7+k2\newQpNWCSP9mo3Ptrzcwbj7/NaOw/8/6z+pcmrrDEy261e+3cIEjbTpcEaT9JOyU6BKlywKRn\n3txrVS380WbmfVWTWHJHr+1qWLpozkUlXsbtQWeqLHw16d7dkfpF1PPtt61NtjWEz7ePbGkr\nVEGSDJipg8ZPNty+4LVE1WP40qZErteHGKaajOrsFowul/6CjlJXpb800Szl/kgd/yhLJ946\nqifepcQw/VEtu9DpAWOgmHmNb2kHjNRLskP7lAfpBBovhybeMTneN6Wi7QteKwtn9EK61+n9\nGqb/EoUuHzDP+CWjgYONVLxfIvdykHIvJ+jlxZh4jcN1duKtHS+pM+mDE2/DIB3kZgNmkTO9\n2OR3XN/OS+14SRxv9nHipThIDBjrm2kHv+zF6RnMCSrXSMlU/PCAyXrRtFDGjbz84sQrv/x9\nGidiztb1015+cOKtmYLOVlBG4VtL57hxkPIvneT6y9+nceJlWSPtHngvGjCGhq5rAbzYJfGS\nKZEt/1tizhQ/uMkOd/PyW+NlyD6VNFHLR4OrtQBe6stfHCRfayQGjKY8XnRb7OHDy2aN1KKJ\nWnyImUvixS6Jl8oSrSuI8CHGXwt4sfHhhSA1Ai9XteDDy7vEMNOgiVqKbpQ/SWGv8ZIohpf6\nEq0riPAhxl8LeLHx4YUgNQIvV7Xgw8vmplUPh+qPiOfmuZTamzPxEhfFi1lieP1Xs5et7uZ1\nImZVEi92yTov1S0U48JLGCQXb7D5ELMqiRe7ZJ2XL59gCNJ+SbzYJfFilnhp4VBtFsWLXRQv\nNSVaVxCzFcHVqTd4sXHhhSA1Ai+XteDCC0FqBF4ua8GFl+Viw8E9/NZz3rkkXuySeEmVOHZp\n8lvFrMrixS6Ll2QJL+9Uv0w8l6cHf+NZhspe4yVRHC9GicwM8/rWVb9eafxMKc/gmZhbzrx4\nSeDBS9EaaZgKGmWaLE9fLp6rx3JuuRbASwoHXor2q0OQHsZDIaJeX3ykxksCB158Bmlc/6WN\nFvVreo2XdC2/5uV4kNrdhPhiEdOk+rsOGLwk6O6lbL+uPlSPf1PL4b/0sc9tBwxeEvT24vHO\nhj96iymt5eIjNV4S9PaylHDzu5w//F2F0b8j8Kai1zkvHY7UeEnQ2csQPaisrK2YVlRc5m3e\nQiV4sens5XyQWvFsVvOLFgrx8sNeKoJUJ+w0Tqo9vsx0sgONqsVLolxp8o43dc9q97381oCZ\ny+NF3fx3inHb0r2qva4lgvQV1V7X0r2qva4lN0Hi1C6xAV7sDfBilhte/x3qzHeKWZXHi10e\nL1a5YfqvWVP3rBYvyfJ4scohJlkeL3Z5vJjlXlquW3HeB7zY4CUAFQACCBKAgODyN7EywIsN\nXgIEFxu+G7zY4CWEIO2AFxu8hBCkHfBig5eQzS1CHXviFbzY4CUAFQACCBKAgHeQKj5Sa1ag\nOMpfdKJQ1V+8JArjxSx34maPvw2HVU1HmeuJXpVS2V+8ZIpf0E6mIrOKfl6Wq3aH+jCMIjFD\n2JvwZQmfTtb2Fy+JzfCSavn4x7RkYpZuDH8/qc/j6RVFC/X9xUtiQ7yYHXp36lh75895h/WD\n4d2TYdmN4cTZRNjCof7iJbEtXqIOeZlhJg1jJOZU9fPS+K4zL16iatb1dfeydKi24Y9HuZj3\nDqzFLK8IWqjtL14Sm+HFLH64WamYYdv/ufZOAwYviY3wYnfo3PsCAjHrg3KDQ/VYLeZdGC/2\nNnhJFD/ZnvINtuDUNLgKI0HW36vawUuiIl9eWnsD+AkIEoAAggQggCABCCBIAAIIEoAAggQg\ngCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAII\nEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCAB\nCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAA\nggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBI\nAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQg\ngCABCCBIAAIIEoAAggQggCABCCBIAAIIEoAAggQggCABCCBIAALmIA1vevbFJXixwUvIx8Rs\nBDUBeLHBy5Zh9e8YPf518GKDlwg0AAgYkk9g/FsJjCwFYl5C8LLmc2o307k7zhg+/zPFhOAl\nIrjYgJYNDBgbvESsLn+jJYIBY4OXiJUJTuwiGDA2sxO8THCxIQsXG2xYUm9BBYCA+WIDM68F\nl3ltJieImVjubEBMDF5sWCNFEKQceLGZRw1eJghSDrzYTAsCvMxsblpFTACnMDbc+x0xWF9g\ngsu8NsPmK2ACQMDq1I6ZNwIvCThQb+FiQw682AwjS8cNBCkHXmy4CBNBkHLgxYYgRcwf7GPA\nGODFhiBFrD6PxBvVFnix+FxpwMsMKgAEbO5sgBDuirfhrvgtyxoJYrjYYIOXiOU+3oNOhnY8\nG9ZdbufggGnYd7zYdPUS3NlQbWVsOic921V9wYCp3qIYvNh09RKskUqSV9/EYf61q/rWA8aH\nl4OXv7/Vy+mLDe3EPB7tzJT2ejh8seHrvZSd8RxuoZ6+Xk7vV0sxj2Z1t18k4+XqFvp6GVYP\nnK2RfAwYvCRL4sUqgZhkSbzYJfFilUBMsiRe7JJ4sUogJlkSL3ZJvFglEJMsiRe7JF6sEohJ\nlsSLXRIvVgnEJEvixS6Jl6oSrStI4mPA+GsBLzZegnTwrepvFbMUrfRy/E3/QvBi4yRI1Tv4\n5WLqS2q22wcvNncN0sntdnk8nu3M3HjA4MWmsxeCJCup2W4XvNi4CZK3NZKXAYMXuyReKku0\nriDF/1Y8DBh3LeDFprOXISpXuaffKiYqhhe7GF7WJeZjdPXB+lvFfMrhxS6Hl1QJznkTJb14\nmT7+ufYi/0joDb1MsEba4GzAeGlh/iD1yov+w9X38zLTeby4CxIDxuaxyFh5UU/B9/NiTbw9\nxovDIDFgLPBiY0+813shSI3AyzUtePFStl/TstIo3VDM6H7A4CVRze95KdqvWcqFYv6+zI99\nDhi8ZGr5MS9LiWH6o1qpQsPmjbg2d397EbMUdeHlgRcbL16G1YMhtcH0ru0FM8x6wDw9DBjn\nXsRibudl9OKlJkjm29gMmG5eGryX/xVe/AZpMXOkiSrmwXGHAYMXB16cTDDBGsnBL0V3N2Dw\nkiqKl7oSrSsIcSPGWQt4sXHjxXOQRgbMxGv/8WLgZrwsayQXd/P6GzC+vajF3M1LKkjXexky\nz0RN1OBvwBwpfnCTDHix8eNlyD6VNFFDIObpZsDgRVT+e70QpCblr/CymoNlfIOXzkHy8TdB\n58Hx+v+5ctJ5LdCyhRLwYhN46OrF2VU7fwPGRQvhTIuXGT/jxVeQGDA2aS9iMXixKQ3SMNOg\niQqic9zOAwYviWJ4qS+R3LLBbfGOxByvGi921d/txfOp3XjLAdOiBbzYOPKyuWnV1aHax1oA\nL4mieDFLDK//el/OdCRmVbK7l810+/ayelHHvbx4Gi9hkDq/wTbt/GeY+Bkwjr1oxeDF5ouC\nlBFzwBgDxgYvNlVrpLH+N6Jvmlh+v+XDePR++gg+B/zYfHP+ui9mfvmxqXPbgYdRpGYtoPSS\nRO4lrrOg2d/1kh+mF1+1+/c4QtD9+etyQmeIOdTOY/2LbC+9OoWXFRsvh3q+7O/8tbsX6eXv\nUgn7YgLD8TeNjfMt/pv/Ee32Lhd6qaozJSfudRsUXhYJSz3j9uFa3wVe9O8j/Us9+meWswwl\nxEQi7TqtDuz2Wo7tJfnPptzGy+q1cfuwzkuy2USv5YQtxAN3z8t7o8/XPS+bbxaNl4NelosN\nBx0K1KcGzHM0B8z5BqsW1R68rF4Yt14yP/56Ong50PP3vu95eT/UjJiqI1Lm7bULfpfzv3H9\nFzr8DBgPXiZCL6IGAm7i5fG3FgrO3cZovBRdbSmk9tQuoWaW0nDABLgaMCNeksV7edkkJOVF\nlSPVEWmY/v3JAYOXRNmeXqLAjLaXy85gitZIw/TlBwcMXhIlO3sJEuLAS9F+TYUCeY3+usAf\nwYqxRQOaXuMlX8sveSnbr2HztbqCShyIqaoGL3Y1P+RF/z6SAgdiXLaAFxsHXir2K3OFRo2D\nc96zdeHlt7x4DdJrLfmcH8m57YDBi4UDL0uJ3aXghWLe12Se8yM5NfMHXuyieDFLHN3BhmfV\nz3ZVV1zmbd5CNXix6erlfJDa8WxYd/n+4QUvBV5OB6nhHNOk4tpK8aLuxHd6Wd3ZUJa84035\nqLh6wODFLo8XdePfKcZxS3i5sGKCdGWl17aElwsrrg8Sh+rEBnixN8CLWW54/XeoK98pZlUe\nL3Z5vFjlhum/Zk35qPjQgMGLVR4vVjnEJMvjxS6PF7PcS8t1K877gBcbvASgAkAAQQIQEFz+\nJlYGeLHBS4DgYsN3gxcbvIQQpB3wYoOXEIK0A15s8BKyuUWoY0+8ghcbvASgAkAAQQIQ8A5S\nxUdqzQo0R/lLThWq+ouXRGG8mOVO3Ozxt+Gwquk4c03Rq0Iq+4uXTPEL2slWZVbSy8ty1e5Q\nD4ZRJmYI+xO+LODTydr+4iWxGV5S7R7/mJZQzNKR4e9n9Xk8vXK+/vr+4iWxIV7M7ry7dKw9\nxTnvsH4wvPsyLDsynDifWNd/qL94SWyLl6g7fmaYScMYiTlR+bw0vu/Mi5eoonWNnb0s3alt\ndlj+TKhYzHsX1mKWV07XX9tfvCQ2w4tZ/HCjYjHDdg/m+rsMGLwkNsKL3Z1z7wtIxKwPyvJD\n9Vgt5l0YL/Y2eEkUP9me9g224OQ0uAojQNjfq9rBS6IqT15aewP4CQgSgACCBCCAIAEIIEgA\nAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCA\nIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggS\ngACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEI\nIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACC\nBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgA\nAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAggSgACCBCCAIAEIIEgAAuYgDW969sUl\neLHBS8jHxGwENQF4scHLlmH17xg9/nXwYoOXCDQACBiST2D8WwmMLAViXkLwsuZzajfTuTvO\nGD7/M8WE4CUiuNiAlg0MGBu8RKwuf6MlggFjg5eIlQlO7CIYMDazE7xMcLEhCxcbbFhSb0EF\ngID5YgMzrwWXeW0mJ4iZWO5sQEwMXmxYI0UQpBx4sZlHDV4mCFIOvNhMCwK8zGxuWkVMAKcw\nNtz7HTFYX2CCy7w2w+YrYAJAwOrUjpk3Ai8JOFBv4WJDDrzYDCNLxw0EKQdebLgIE0GQcuDF\nhiBFzB/sY8AY4MWGIEWsPo/EG9UWeLH4XGnAywwqAARs7myAEO6Kt+Gu+C3LGgliuNhgg5eI\n5T7eg06Gdjwb1l1u5+CAadh3vH1leocAAArVSURBVNh09RLc2VBtZWw6Jz3bVS0aMJkznJ/2\ncr6FI3T1EqyRSpJX34SUf5pqNAMm9+2v93Lw8ve3ejl9sUElpmyPHw+NmdJeD9mLDVcEya2X\nsjOewy3s4c3L6f0SiSnc48fjIWlO0+sLgnRLLxe04M7LsHqQLm2tBY5PShbWHhvnvNcPmJyX\n3BnOL3vRtJDFnZeiIF0y8zoTsyqZmWDSk8lPe5G0kMWdF4K0X9L/gHncwMuXn8HcK0h3GDDV\nLWS5p5f2bwvcN0gu1gLeBkymul/28osTb+HFhi5rAbOUpwGjaSFL2R4789IpSGYpX0E61UQR\n7sSsSroP0v9FXHn56SC1a6IItzNvdWXyRbXPIO3W0nwp4Gu8LCVKfvzXionOeV9FNGYqep3x\nkrun8Ze9XLMU2PfS69SuURMlmHOqiwGTLZn5kww/7aXBdhvcTTAE6VzJ5AfRL/TyV8KZF/l2\nIf7GS02QGq4F/IkpLplQIPMS7/F9gtTslNffeFmC1HUtYImJ70vsEqT96eI3g3RwGv36IOVL\n9VgLRGfBXQZM3xaKhsJveim7mnlxkIbolahc87VAtMsOgrTnJfNxpeuC9P6+Ky/Wd9ovBaxS\nVx+R5v1L7+jlawEHQdrx0v6NR6dB2vNyxVLA2ZG6aI2Uq+27g7Szdhy7BOkZl/Hl5ZKlgLO1\n4+n9ujpIGjOaXl8TpO0eby7CfJ46WyN1WArcIUhd1gLbIH2eOgtSqrqWQXpEReJSxxBebPix\ntwWK3w5oOvOau+w/SPMKs1UL5jH4JkFyckS6ygtBagRBatiCfawx144/GKRxO0DuGSTlZd67\nBqn1UsAO0j+jkL8gRaVbD5jtWwXTGtvTgGndwuRlc3XBe5C6TLxbTd2CNEx/VCtd6tK1QByk\n8OspKnqd86JpIY05FpwEKeOlV5B6nsGsDzapK5YnmyjADtLTQ5AceMkGaX7iyAtBIkhWyd5e\ntkfmwIvXIKWqu3DteOlS4DZB8jhgNC1kuGmQzKXAl1+ECdZI3X4pek2QcmaKndWtBTp7KQuS\nZMTcyovTIDVrYp95vPwLXqweMOWf0L/VVbtwl71MMAchSKUVHOi3uc+bq3YFA+ZR/rsuOgyY\nE162IphgjLcFtj/8XkHSfOLxMVFVhz3xBs/jh9Gf9XhsL5h/XrZarFgLiD4JupJSLKcgSPbD\nuOnE97YvX+/lEPYpr/12yUWnvEPmWXUTD4vUhqvvWvv899CcbqfpaFt1sMVOVyp39PRaoEzK\n9mVNkN4TjPFNoydXezmG6AymgupTu1MzzL/VpFtK+BGAcFg894IUDMv55XTry0Gsdj9PzryP\nVLfiQuEr2wfjdo1UFKRxObvLd+FqLwdJBKnrRZjjQTIuZ27Otj4hSfH+7vLvWBkke5pfP1n6\n82/TvWsHzDJc/80vREPZGNbW7u95We1lcLhP/AQ2bf5YkIQXYTZrpNKKa5oY1+N4M6K3l19S\nAyZStB414dQaT+xHe/0uWeslnmD+WR0yZplV37fvJZYHaZmizKAuE8wq06vOV+5mcfF0C1FH\n9yk4UhcEyZi4jEdv3F/+PhukhX/GozTXXp3K9yg8Qn/YfOq1LkjWAS/dizOnvPUcWVMXrh2r\nghQ3F/8Q7F7b9H4faZOKkiCN77n7VLPuLvPOYzmM1fLq6lFugllOebcVF7R9sZd/m2OkHSkj\nXqmlQPnacarzsRYbUTPBvEvkfu1LUQWH2R5eQkfJIJ1dQBb2es9L5ruHvayORYkgZbzEC6U6\nRF4KW7B6aR6lomPH9kFmRg4nm6WR+dH6ldQcdI8jUtqRuaq8MEiltehvzjw4wTg+UtdELzhA\nrR9Po/5TriZIiVRGL5lzkP8gbVMRRue3g7Q/wfQ6Uu9WIztSr69LzWdb+0fqvbXjqoX4lUO9\n3ty0ev3lzO1PfzXpPINn6yJnx0v9Zf58LdcGKbN2vDBIXbyUrh1jd6sXrHdpdqi5/P36r8M7\n1d6DlPPSMEi5I/UzfJra5AjevSTPeGuCdIDaIPV4gy060s5H7+cyA21/acHFQUqOmAYXGz4U\nBan3BJMof92ROhWk69eO7oK0Ono/10/CLc42KgqSpIUE0QQzfyFIwdP5gY8g/WlJlbd+vZLq\nE4/Og5T1omnB5lCQzjYq8tL7SN3hlLdov4apYIO1QPTzD7Jjnt7Gf4GsGvFl3hZe0kGKvixb\nnG3U3RvVEZkgxU/NTY5wxyAF4bHXiadzdIMBs/VSFCRvE0yDFg4F6YJTXhdBMn6xXzZI5/E/\nYA4Eyd8E0/5IbQepw9pxWRdmyrYNkvWrZv/2/ClZJ5pULKprJyPV2nGMPkn9/pK+milA6uVk\nCwnSQeq6dlyVyPzwrYsNxU1UMx29n6vHYmp6nfaSu+VMPsEsB+fn2OxQrfGiasHG50WY7Wza\noIlqnAVpTHtp/pfp1qyP1K0O1SIvVmXCI3UcpPUZzBgH6YK1Y9kR6VQT1YRBapEj3czb+C/T\nrdkGqYWZOxypt6e887/P8OnCFWvHojVSrr4Wy/ZPelwEaddLYjx9+5E666XtkXpzm8v87+qU\n9/q14+n9ahikzUMl0l5fHaT3Y99eLjtST58jmp9K1kRbviJIDRq4xeXvmFtNMFcdqR/BcrHb\nKW/ZfrW8RcjiLkG69mrm3SaYy9aOmwmmyymvgzdkDYI1UosGNL3Gi4sWpvB0XVN7DdJrtfgc\nW6wb/7htkCYvgitRFvc8UgfjpdMEs5TY+cTjpQPmPUx8DBinXtpwfy+9xssQPags1PBkwMWA\nwUt9yeuD9EdXL0VBmr5pHqrb8WxYt0YhXpLfSwapHV29lAWprAo1TSqurRQv9eXzg+o7vSz7\nnEvekF48fquYuXzhjHS+pb4VS71kjtTf6uX0VbtvFeO4JbxcWDFBurLSa1vCy4UV1wcpc6j+\n6SBxapfYAC9mueH1X/6U9yeDlPOibalvxXg5V2kYpJ0k/dLicVU+7UXaUt+K8XKu0rIgSZry\nUTEDRlMpXlLlXlquW3HeB7zY4CUAFQACCBKAgODyN7EywIsNXgIEFxu+G7zY4CWEIO2AFxu8\nhBCkHfBig5eQzS1CHXviFbzY4CUAFQACCBKAgHeQKj5Sa1agOcpfcqpQ1V+8JArjxSx34maP\nvw2HVU3HmWuKXhVS2V+8ZIpf0E62KrOSXl6Wq3aHejCMMjFD2J/wZQGfTtb2Fy+JzfCSavf4\nx7SEYpaODO9fxP5+PL1yvv76/uIlsSFezO68u3SsPcU577B+MIzT3zMYwldO13+ov3hJbIuX\nqDt+ZphJwxiJOVH5vDS+78yLl6iidY2dvSzdqW12WP6alFjMexfWYpZXTtdf21+8JDbDi1n8\ncKNiMcN2D+b6uwwYvCQ2wovdnXPvC0jErA/K8kP1WC3mXRgv9jZ4SRQ/2Z72Dbbg5DS4CiNA\n2N+r2sFLoipPXlp7A/gJCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAg\nAQggSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQggSAACCBKA\nAIIEIIAgAQggSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQgg\nSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQggSAACCBKAAIIEIIAgAQj4D/Nt1r8K5UeV\nAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot where each row is shape type\n",
    "# each col is transcription factor\n",
    "plot.new()\n",
    "layout(matrix(1:12, ncol=length(shapes)))\n",
    "for (n in ns) {\n",
    "    plotShape(shapes[[n]]$MGW)\n",
    "    plotShape(shapes[[n]]$ProT)\n",
    "    plotShape(shapes[[n]]$Roll)\n",
    "    plotShape(shapes[[n]]$HelT)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They all seems to have the same pattern of MGW, ProT, Roll, and HelT along the sequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Write an R function that takes as input a FASTA file. The function returns a feature matrix for “1-mer” sequence model using one-hot encoding which is a way to represent categorical data as binary vector [3pt].\n",
    "\n",
    "For DNA, we have four categories A, T, G, and C. Thus a one hot code for DNA is: <br>\n",
    "A: [0,0,0,1] <br>\n",
    "C: [0,0,1,0] <br>\n",
    "G: [0,1,0,0] <br>\n",
    "T: [1,0,0,0] <br>\n",
    "\n",
    "For example, for the sequence AATTC, the 1-mer one-hot encoding is: <br>\n",
    "[0,0,0,1,0,0,0,1,1,0,0,0,1,0,0,0,1,0,0,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# encode nucleotide\n",
    "# gets nucleotide and returns the corresponding one-hot encoding\n",
    "encodeNucleotide <- function(nucleotide) {\n",
    "    dict <- list(A=c(0,0,0,1), C=c(0,0,1,0), G=c(0,1,0,0), T=c(1,0,0,0))\n",
    "    return(dict[nucleotide])\n",
    "}\n",
    "\n",
    "# encode sequence\n",
    "# gets sequence and returns vector of one-hot encoded sequence\n",
    "encodeSequence <- function(sequence) {\n",
    "    nucleotides <- strsplit(sequence, split='', fixed=T)\n",
    "    result <- unlist(lapply(nucleotides, encodeNucleotide))\n",
    "    names(result) <- c()\n",
    "    return(result)\n",
    "}\n",
    "\n",
    "# encode FASTA\n",
    "# gets FASTA file and returns matrix of one-hot encoded sequences\n",
    "encodeFASTA <- function(filename) {\n",
    "    # read FASTA\n",
    "    fa <- read.table(filename, comment.char='>')\n",
    "    # encode sequence\n",
    "    result <- apply(fa, 1, encodeSequence)\n",
    "    # transpose\n",
    "    result <- t(result)\n",
    "    return(result)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TEST**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "$A\n",
      "[1] 0 0 0 1\n",
      "\n",
      "$C\n",
      "[1] 0 0 1 0\n",
      "\n",
      "$G\n",
      "[1] 0 1 0 0\n",
      "\n",
      "$T\n",
      "[1] 1 0 0 0\n",
      "\n",
      " [1] 0 0 0 1 0 0 0 1 1 0 0 0 1 0 0 0 0 0 1 0\n",
      "[1] 7534  144\n",
      "[1] TRUE\n",
      "[1] TRUE\n"
     ]
    }
   ],
   "source": [
    "nucleotides <- c('A', 'C', 'G', 'T')\n",
    "for (i in nucleotides) {\n",
    "    print(encodeNucleotide(i))\n",
    "}\n",
    "\n",
    "sequence <- 'AATTC'\n",
    "print(encodeSequence(sequence))\n",
    "\n",
    "sequence <- 'GGGCATGAAAGCCACCTCGTGGTTTTGCAGCAAGTG'\n",
    "len <- length(unlist(strsplit(sequence, split='', fixed=T)))\n",
    "test <- encodeFASTA('Mad.fa')\n",
    "print(dim(test))\n",
    "print(dim(test)[2] == len*4)\n",
    "print(all(test[1,] == encodeSequence(sequence)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3. Use the function from Q#2 to generate a “1-mer” feature vector and combine the shape features generated from Q#1 for “1-mer+shape” model with respect to the datasets of Mad, Max and Myc. Use the caret package to build L2-regularized MLR models for “1-mer” and “1-mer+shape” features with 10-fold cross validation (set the L2 lambda value between 2-15 to 215), and print out the average R2 (coefficient of determination) for these two models with respect to the datasets of Mad, Max and Myc. The corresponding protein–DNA binding affinities can be found in corresponding `*.s` files. Compare the performance between “1-mer” sequence model and “1-mer+shape” model and explain what you find [3pt]."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create list of feature matrices for each file\n",
    "features.seq <- list()\n",
    "features.shape <- list()\n",
    "\n",
    "# loop over files\n",
    "for (n in ns) {\n",
    "    # add 1-mer feature vector\n",
    "    features.seq[[n]] <- encodeFASTA(paste(n, '.fa', sep=''))\n",
    "    # combine intra bp parameters\n",
    "    features.shape[[n]] <- cbind(features.seq[[n]], shapes[[n]]$MGW, shapes[[n]]$ProT)\n",
    "    # combine inter bp parameters\n",
    "    features.shape[[n]] <- cbind(features.shape[[n]], shapes[[n]]$Roll, shapes[[n]]$HelT)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove columns with NA\n",
    "for (n in ns) {\n",
    "    features.seq[[n]] <- features.seq[[n]][, colSums(is.na(features.seq[[n]])) == 0]\n",
    "    features.shape[[n]] <- features.shape[[n]][, colSums(is.na(features.shape[[n]])) == 0]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add ground truth to features\n",
    "data.seq <- list()\n",
    "data.shape <- list()\n",
    "for (n in ns) {\n",
    "    gt <- read.table(paste(n, '.s', sep=''))[,2]\n",
    "    data.seq[[n]] <- cbind(features.seq[[n]], gt)\n",
    "    data.shape[[n]] <- cbind(features.shape[[n]], gt)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Shape Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split data\n",
    "train.data.shape <- list()\n",
    "test.data.shape <- list()\n",
    "for (n in ns) {\n",
    "    trainIndex <- createDataPartition(data.shape[[n]][,ncol(data.shape[[n]])], p=.8, list = FALSE, times=1)\n",
    "    train.data.shape[[n]] <- data.shape[[n]][trainIndex,]\n",
    "    test.data.shape[[n]] <- data.shape[[n]][-trainIndex,]\n",
    "}\n",
    "\n",
    "# 10-fold cross validation\n",
    "fitControl <- trainControl(method=\"cv\", number=10, savePredictions=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo, :\n",
      "\"There were missing values in resampled performance measures.\"Warning message in nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo, :\n",
      "\"There were missing values in resampled performance measures.\"Warning message in nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo, :\n",
      "\"There were missing values in resampled performance measures.\""
     ]
    }
   ],
   "source": [
    "# train\n",
    "model.shape <- list()\n",
    "for (n in ns) {\n",
    "    model.shape[[n]] <- train(gt~., data=data.frame(train.data.shape[[n]]), trControl=fitControl, method=\"glmnet\",\n",
    "                   tuneGrid=data.frame(alpha = 0, lambda=c(2^c(-15:15))))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sequence Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split data\n",
    "train.data.seq <- list()\n",
    "test.data.seq <- list()\n",
    "for (n in ns) {\n",
    "    trainIndex <- createDataPartition(data.seq[[n]][,ncol(data.seq[[n]])], p=.8, list = FALSE, times=1)\n",
    "    train.data.seq[[n]] <- data.seq[[n]][trainIndex,]\n",
    "    test.data.seq[[n]] <- data.seq[[n]][-trainIndex,]\n",
    "}\n",
    "\n",
    "# cross validation\n",
    "fitControl <- trainControl(method=\"cv\", number=10, savePredictions=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo, :\n",
      "\"There were missing values in resampled performance measures.\"Warning message in nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo, :\n",
      "\"There were missing values in resampled performance measures.\"Warning message in nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo, :\n",
      "\"There were missing values in resampled performance measures.\""
     ]
    }
   ],
   "source": [
    "# train\n",
    "model.seq <- list()\n",
    "for (n in ns) {\n",
    "    model.seq[[n]] <- train(gt~., data=data.frame(train.data.seq[[n]]), trControl=fitControl, method=\"glmnet\",\n",
    "                   tuneGrid=data.frame(alpha = 0, lambda=c(2^c(-15:15))))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Average R2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Mad\"\n",
      "[1] \"Shape model: 0.73535729130494\"\n",
      "[1] \"Sequence model: 0.711448347885396\"\n",
      "[1] \"Max\"\n",
      "[1] \"Shape model: 0.719817550912976\"\n",
      "[1] \"Sequence model: 0.702892352169605\"\n",
      "[1] \"Myc\"\n",
      "[1] \"Shape model: 0.729568107763171\"\n",
      "[1] \"Sequence model: 0.72835286320677\"\n"
     ]
    }
   ],
   "source": [
    "for (n in ns) {\n",
    "    print(n)\n",
    "    print(paste('Shape model:', mean(model.shape[[n]]$results$Rsquared[!is.na(model.shape[[n]]$results$Rsquared)])))\n",
    "    print(paste('Sequence model:', mean(model.seq[[n]]$results$Rsquared[!is.na(model.seq[[n]]$results$Rsquared)])))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Mad\"\n",
      "[1] \"Shape model RMSE: 0.306371205449735\"\n",
      "[1] \"Sequence model RMSE: 0.371845394578494\"\n",
      "[1] \"Max\"\n",
      "[1] \"Shape model RMSE: 0.0794624565461406\"\n",
      "[1] \"Sequence model RMSE: 0.100545693240345\"\n",
      "[1] \"Myc\"\n",
      "[1] \"Shape model RMSE: 0.309649757237914\"\n",
      "[1] \"Sequence model RMSE: 0.374833561133695\"\n"
     ]
    }
   ],
   "source": [
    "for (n in ns) {\n",
    "    print(n)\n",
    "    pred.shape <- predict(model.shape[[n]], test.data.shape[[n]])\n",
    "    pred.seq <- predict(model.seq[[n]], test.data.seq[[n]])\n",
    "    print(paste('Shape model RMSE:', RMSE(pred.shape, data.frame(test.data.shape[[n]])$gt)))\n",
    "    print(paste('Sequence model RMSE:', RMSE(pred.seq, data.frame(test.data.seq[[n]])$gt)))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems the model that takes shape as input is better at predicting binding affinity than the model that only takes in the sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TEST**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 7534  286\n",
      "logical(0)\n",
      "[1] TRUE\n",
      "[1] 8568  286\n",
      "logical(0)\n",
      "[1] TRUE\n",
      "[1] 6926  286\n",
      "logical(0)\n",
      "[1] TRUE\n"
     ]
    }
   ],
   "source": [
    "# test for correct dimensions\n",
    "\n",
    "# MGW, ProT => intra bp parameter (number of features per sequence = lengthOfSequence)\n",
    "# Roll, HelT => inter bp parameter (number of features per sequence = lengthOfSequence - 1)\n",
    "len <- 36\n",
    "for (n in ns) {\n",
    "    print(dim(features.shape[[n]]))\n",
    "    print(dim(features.shape[[n]])[1] == dim(shapes[[n]][1])[1])\n",
    "    x <- len*4 + len*2 + (len-1)*2\n",
    "    print(dim(features.shape[[n]])[2] == x)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] TRUE\n"
     ]
    }
   ],
   "source": [
    "# test for NA\n",
    "for (n in ns) {\n",
    "    print(!any(is.na(features.shape[[n]])))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] TRUE\n"
     ]
    }
   ],
   "source": [
    "# test dimensions for data\n",
    "for (n in ns) {\n",
    "    print(dim(features.shape[[n]])[1] == dim(data.shape[[n]])[1])\n",
    "    print(dim(features.shape[[n]])[2] == dim(data.shape[[n]])[2]-1)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] \"6029 6027.2\"\n",
      "[1] \"1505 1506.8\"\n",
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] \"6856 6854.4\"\n",
      "[1] \"1712 1713.6\"\n",
      "[1] TRUE\n",
      "[1] TRUE\n",
      "[1] \"5542 5540.8\"\n",
      "[1] \"1384 1385.2\"\n"
     ]
    }
   ],
   "source": [
    "# test dimensions of test & training data\n",
    "for (n in ns) {\n",
    "    print(dim(train.data.shape[[n]])[2] == dim(data.shape[[n]])[2])\n",
    "    print(dim(test.data.shape[[n]])[2] == dim(data.shape[[n]])[2])\n",
    "    print(paste(dim(train.data.shape[[n]])[1], 0.8*dim(data.shape[[n]])[1]))\n",
    "    print(paste(dim(test.data.shape[[n]])[1], 0.2*dim(data.shape[[n]])[1]))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
